### PR TITLE
Add E2EDiagnosticSetting structure and api to get/set E2EDiagnosticSetting in registry

### PR DESCRIPTION
**Pull request by michaeljqzq**
# Checklist
- [ ] I have read the [contribution guidelines] (https://github.com/Azure/azure-iot-sdk-c/blob/master/.github/CONTRIBUTING.md).
- [ ] I added or modified the existing tests to cover the change (we do not allow our test coverage to go down).
- If this is a modification that impacts the behavior of a public API
  - [ ] I edited the corresponding document in the `devdoc` folder and added or modified requirements.
- I submitted this PR against the correct branch: 
  - [ ] This pull-request is submitted against the `master` branch. 

# Reference/Link to the issue solved with this PR (if any)

# Description of the problem

# Description of the solution

### CODE REVIEW

File: service/Microsoft.Azure.Devices/RegistryManager.cs [hunk 1]
```diff
@@ -634,5 +634,33 @@ public void Dispose()
...
         /// <returns>Updated Twin instance</returns>
         public abstract Task<Twin> ReplaceTwinAsync(string deviceId, string newTwinJson, string etag, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Update the diagnostic settings
+        /// </summary>
+        /// <param name="deviceId">Device Id</param>
+        /// <param name="setting"><see cref="E2EDiagnosticSetting"/></param>
+        /// <param name="etag">Device ETag</param>
+        /// <returns>echoes back the E2EDiagnosticSetting object</returns>
```
Comment (SLdragon) [inline service/Microsoft.Azure.Devices/RegistryManager.cs]: echoes back the E2EDiagnosticSetting object
-->Echo

File: service/Microsoft.Azure.Devices/RegistryManager.cs [hunk 2]
```diff
@@ -634,5 +634,33 @@ public void Dispose()
...
+        public abstract Task<E2EDiagnosticSetting> UpdateE2EDiagnosticSettingAsync(string deviceId, E2EDiagnosticSetting setting, string etag);
+
+        /// <summary>
+        /// Update the diagnostic settings
+        /// </summary>
+        /// <param name="deviceId">Device Id</param>
+        /// <param name="setting"><see cref="E2EDiagnosticSetting"/></param>
+        /// <param name="etag">Device ETag</param>
+        /// <param name="cancellationToken">
+        /// The token which allows the the operation to be cancelled.
```
Comment (SLdragon) [inline service/Microsoft.Azure.Devices/RegistryManager.cs]: The token which allows the **the** operation to be cancelled

File: service/Microsoft.Azure.Devices/RegistryManager.cs [hunk 3]
```diff
@@ -634,5 +634,33 @@ public void Dispose()
...
+        /// </param>
+        /// <returns>echoes back the E2EDiagnosticSetting object</returns>
```
Comment (SLdragon) [inline service/Microsoft.Azure.Devices/RegistryManager.cs]: Echo

File: service/Microsoft.Azure.Devices/RegistryManager.cs [hunk 4]
```diff
@@ -634,5 +634,33 @@ public void Dispose()
...
+        public abstract Task<E2EDiagnosticSetting> UpdateE2EDiagnosticSettingAsync(string deviceId, E2EDiagnosticSetting setting, string etag, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Get the diagnostic settings
+        /// </summary>
+        /// <param name="deviceId">Device Id</param>
+        /// <returns>the desired E2EDiagnosticSetting object</returns>
```
Comment (SLdragon) [inline service/Microsoft.Azure.Devices/RegistryManager.cs]: the desired E2EDiagnosticSetting object
-->
Desired E2EDiagnosticSetting object

File: service/Microsoft.Azure.Devices/HttpRegistryManager.cs [hunk 5]
```diff
@@ -740,6 +740,58 @@ public override Task<Twin> ReplaceTwinAsync(string deviceId, string newTwinJson,
             return this.ReplaceTwinAsync(deviceId, twin, etag, cancellationToken);
         }
 
+        public override Task<E2EDiagnosticSetting> UpdateE2EDiagnosticSettingAsync(string deviceId, E2EDiagnosticSetting setting, string etag)
+        {
+            return this.UpdateE2EDiagnosticSettingAsync(deviceId, setting, etag, CancellationToken.None);
+        }
+
+        public override async Task<E2EDiagnosticSetting> UpdateE2EDiagnosticSettingAsync(string deviceId, E2EDiagnosticSetting setting, string etag, CancellationToken cancellationToken)
```
Comment (SLdragon) [inline service/Microsoft.Azure.Devices/HttpRegistryManager.cs:748]: cancellationToken defined but not used?

Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/HttpRegistryManager.cs:748]: fixed

File: service/Microsoft.Azure.Devices/RegistryManager.cs (hunk 1 above)
Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/RegistryManager.cs]: fixed

File: service/Microsoft.Azure.Devices/RegistryManager.cs (hunk 2 above)
Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/RegistryManager.cs]: fixed

File: service/Microsoft.Azure.Devices/RegistryManager.cs (hunk 3 above)
Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/RegistryManager.cs]: fixed

File: service/Microsoft.Azure.Devices/RegistryManager.cs (hunk 4 above)
Comment (michaeljqzq) [reply service/Microsoft.Azure.Devices/RegistryManager.cs]: fixed

**Pull request closed without merging**